Source/JavaScriptCore/jit/RegisterSet.h

```diff
@@ -27,183 +27,500 @@
+    JS_EXPORT_PRIVATE static RegisterSet ftlCalleeSaveRegisters(); // Registers that might be saved and used by the FTL JIT.
+    JS_EXPORT_PRIVATE static RegisterSet stubUnavailableRegisters(); // The union of callee saves and special registers.
     JS_EXPORT_PRIVATE static RegisterSet argumentGPRS();
+    JS_EXPORT_PRIVATE static RegisterSetBuilder registersToSaveForJSCall(RegisterSetBuilder live);
+    JS_EXPORT_PRIVATE static RegisterSetBuilder registersToSaveForCCall(RegisterSetBuilder live);
+};
+
+class RegisterSet final {
+    friend ScalarRegisterSet;
+    friend RegisterSetBuilder;
 
-    static RegisterSet registersToNotSaveForJSCall();
-    static RegisterSet registersToNotSaveForCCall();
-    
-    void set(Reg reg, bool value = true)
+    constexpr explicit inline RegisterSet(const RegisterSetBuilder& set)
+        : m_bits(set.m_bits)
+        , m_upperBits(set.m_upperBits)
     {
-        ASSERT(!!reg);
-        m_bits.set(reg.index(), value);
+        m_bits.merge(m_upperBits);
     }
-    
-    void set(JSValueRegs regs, bool value = true)
+
+public:
+    constexpr RegisterSet() = default;
+
+    inline constexpr bool contains(Reg reg, Width width) const
     {
-        if (regs.tagGPR() != InvalidGPRReg)
-            set(regs.tagGPR(), value);
-        set(regs.payloadGPR(), value);
+        ASSERT(m_bits.count() >= m_upperBits.count());
+        if (LIKELY(width < conservativeWidth(reg)) || width <= conservativeWidthWithoutVectors(reg))
+            return m_bits.get(reg.index());
+        return m_bits.get(reg.index()) && m_upperBits.get(reg.index());
     }
 
-    void set(const RegisterSet& other, bool value = true) { value ? merge(other) : exclude(other); }
+    inline constexpr bool contains(Reg reg, IgnoreVectorsTag) const
+    {
+        return contains(reg, conservativeWidthWithoutVectors(reg));
+    }
 
-    void clear(Reg reg)
+    inline size_t numberOfSetGPRs() const
     {
-        ASSERT(!!reg);
-        set(reg, false);
+        RegisterBitmap temp = m_bits;
+        temp.filter(RegisterSetBuilder::allGPRs().m_bits);
+        return temp.count();
     }
-    
-    bool get(Reg reg) const
+
+    inline size_t numberOfSetFPRs() const
     {
-        ASSERT(!!reg);
-        return m_bits.get(reg.index());
+        RegisterBitmap temp = m_bits;
+        temp.filter(RegisterSetBuilder::allFPRs().m_bits);
+        return temp.count();
     }
-    
-    template<typename Iterable>
-    void setAll(const Iterable& iterable)
+
+    inline constexpr size_t numberOfSetRegisters() const
     {
-        for (Reg reg : iterable)
-            set(reg);
+        ASSERT(m_bits.count() >= m_upperBits.count());
+        return m_bits.count();
     }
-    
-    // Also allow add/remove/contains terminology, which means the same thing as set/clear/get.
-    bool add(Reg reg)
+
+    inline size_t sizeOfSetRegisters() const
```

Comment:
Rename this to `byteSizeOfSetRegisters` since we usually use `size` for the # of items.